File: Module.Persist.TPM/Model/History/HistoricalPromo.cs
using Core.History;
using Module.Persist.TPM.Model.TPM;
using System;
namespace Module.Persist.TPM.Model.History {

    [AssociatedWith(typeof(Promo))]
    public class HistoricalPromo : BaseHistoricalEntity<System.Guid>
    {
        public DateTimeOffset? StartDate { get; set; }
        public DateTimeOffset? EndDate { get; set; }
        public DateTimeOffset? DispatchesStart { get; set; }
        public DateTimeOffset? DispatchesEnd { get; set; }
        public DateTimeOffset? LastChangedDate { get; set; }
        public DateTimeOffset? LastChangedDateDemand { get; set; }
        public DateTimeOffset? LastChangedDateFinance { get; set; }
        public int? PromoDuration { get; set; }
        public int? DispatchDuration { get; set; }
        public string InvoiceNumber { get; set; }
        public string DocumentNumber { get; set; }
        public string ClientHierarchy { get; set; }
        public string ProductHierarchy { get; set; }
        public int? Number { get; set; }
        public string Name { get; set; }
        public string OtherEventName { get; set; }
        public string EventName { get; set; }
        public string Comment { get; set; }
        public string CreatorLogin { get; set; }

        //event
        public string PromoEventName { get; set; }
        public string PromoEventDescription { get; set; }

        // Calculation
        /*public int? ShopperTi { get; set; }
        public int? MarketingTi { get; set; }
        public int? Branding { get; set; }
        public int? TotalCost { get; set; }
        public int? BTL { get; set; }
        public double? CostProduction { get; set; }
        public double? PlanUplift { get; set; }
        public int? PlanIncrementalLsv { get; set; }
        public int? PlanTotalPromoLsv { get; set; }
        public int? PlanPromoPostPromoEffectLSV { get; set; }
        public int? PlanRoi { get; set; }
        public int? PlanIncrementalNsv { get; set; }
        public int? PlanTotalPromoNsv { get; set; }
        public int? PlanIncrementalMac { get; set; }
        public double? PlanPromoXSites { get; set; }
        public double? PlanPromoCatalogue { get; set; }
        public double? PlanPromoPOSMInClient { get; set; }
        public double? PlanPromoCostProdXSites { get; set; }
        public double? PlanPromoCostProdCatalogue { get; set; }
        public double? PlanPromoCostProdPOSMInClient { get; set; }
        public double? ActualPromoXSites { get; set; }
        public double? ActualPromoCatalogue { get; set; }
        public double? ActualPromoPOSMInClient { get; set; }
        public double? ActualPromoCostProdXSites { get; set; }
        public double? ActualPromoCostProdCatalogue { get; set; }
        public double? ActualPromoCostProdPOSMInClient { get; set; }
        public double? PlanPromoBaselineLSV { get; set; }
        public double? PlanPromoIncrementalBaseTI { get; set; }
        public double? PlanPromoIncrementalCOGS { get; set; }
        public double? PlanPromoTotalCost { get; set; }
        public double? PlanPromoNetIncrementalLSV { get; set; }
        public double? PlanPromoNetLSV { get; set; }
        public double? PlanPromoNetIncrementalMAC { get; set; }
        public double? PlanPromoIncrementalEarnings { get; set; }
        public double? PlanPromoNetIncrementalEarnings { get; set; }
        public int? PlanPromoROIPercent { get; set; }
        public int? PlanPromoNetUpliftPercent { get; set; }
        public double? ActualPromoBaselineLSV { get; set; }
        public int? ActualInStoreDiscount { get; set; }
        public double? ActualInStoreShelfPrice { get; set; }
        public double? ActualPromoIncrementalBaseTI { get; set; }
        public double? ActualPromoIncrementalCOGS { get; set; }
        public double? ActualPromoTotalCost { get; set; }
        public double? ActualPromoNetIncrementalLSV { get; set; }
        public double? ActualPromoNetLSV { get; set; }
        public double? ActualPromoNetIncrementalMAC { get; set; }
        public double? ActualPromoIncrementalEarnings { get; set; }
        public double? ActualPromoNetIncrementalEarnings { get; set; }
        public int? ActualPromoNetROIPercent { get; set; }
        public int? ActualPromoNetUpliftPercent { get; set; }
        public double? PlanPromoBaselineBaseTI { get; set; }
        public double? PlanPromoBaseTI { get; set; }
        public double? PlanPromoNetNSV { get; set; }
        public double? ActualPromoBaselineBaseTI { get; set; }
        public double? ActualPromoBaseTI { get; set; }
        public double? ActualPromoNetNSV { get; set; }

        // Promo Closure
        public int? FactShopperTi { get; set; }
        public double? FactMarketingTi { get; set; }
        public int? FactBranding { get; set; }
        public int? FactBTL { get; set; }
        public double? FactCostProduction { get; set; }
        public int? FactTotalCost { get; set; }
        public double? FactUplift { get; set; }
        public int? FactIncrementalLsv { get; set; }
        public int? FactTotalPromoLsv { get; set; }
        public int? ActualPromoPostPromoEffectLSV { get; set; }
        public int? FactRoi { get; set; }
        public int? FactIncrementalNsv { get; set; }
        public int? FactTotalPromoNsv { get; set; }
        public int? FactIncrementalMac { get; set; }*/

        // Calculation
        public double? PlanPromoTIShopper { get; set; }
        public double? PlanPromoTIMarketing { get; set; }
        public double? PlanPromoBranding { get; set; }
        public double? PlanPromoCost { get; set; }
        public double? PlanPromoBTL { get; set; }
        public double? PlanPromoCostProduction { get; set; }
        public double? PlanPromoUpliftPercent { get; set; }
        public double? PlanPromoIncrementalLSV { get; set; }
        public double? PlanPromoLSV { get; set; }

        //необходимость полей в таком виде под вопросом
        public double? PlanPromoPostPromoEffectLSV { get; set; }
        public double? PlanPromoPostPromoEffectLSVW1 { get; set; }
        public double? PlanPromoPostPromoEffectLSVW2 { get; set; }
        //

        public double? PlanPromoROIPercent { get; set; }
        public double? PlanPromoIncrementalNSV { get; set; }
        public double? PlanPromoNetIncrementalNSV { get; set; }
        public double? PlanPromoIncrementalMAC { get; set; }
        public double? PlanPromoXSites { get; set; }
        public double? PlanPromoCatalogue { get; set; }
        public double? PlanPromoPOSMInClient { get; set; }
        public double? PlanPromoCostProdXSites { get; set; }
        public double? PlanPromoCostProdCatalogue { get; set; }
        public double? PlanPromoCostProdPOSMInClient { get; set; }
        public double? ActualPromoXSites { get; set; }
        public double? ActualPromoCatalogue { get; set; }
        public double? ActualPromoPOSMInClient { get; set; }
        public double? ActualPromoCostProdXSites { get; set; }
        public double? ActualPromoCostProdCatalogue { get; set; }
        public double? ActualPromoCostProdPOSMInClient { get; set; }
        public double? PlanPromoBaselineLSV { get; set; }
        public double? PlanPromoIncrementalBaseTI { get; set; }
        public double? PlanPromoIncrementalCOGS { get; set; }
        public double? PlanPromoTotalCost { get; set; }
        public double? PlanPromoNetIncrementalLSV { get; set; }
        public double? PlanPromoNetLSV { get; set; }
        public double? PlanPromoNetIncrementalMAC { get; set; }
        public double? PlanPromoIncrementalEarnings { get; set; }
        public double? PlanPromoNetIncrementalEarnings { get; set; }
        public double? PlanPromoNetROIPercent { get; set; }
        public double? PlanPromoNetUpliftPercent { get; set; }
        public double? ActualPromoBaselineLSV { get; set; }
        public double? ActualInStoreDiscount { get; set; }
        public double? ActualInStoreShelfPrice { get; set; }
        public double? PlanInStoreShelfPrice { get; set; }
        public double? ActualPromoIncrementalBaseTI { get; set; }
        public double? ActualPromoIncrementalCOGS { get; set; }
        public double? ActualPromoTotalCost { get; set; }
        public double? ActualPromoNetIncrementalLSV { get; set; }
        public double? ActualPromoNetLSV { get; set; }
        public double? ActualPromoNetIncrementalMAC { get; set; }
        public double? ActualPromoIncrementalEarnings { get; set; }
        public double? ActualPromoNetIncrementalEarnings { get; set; }
        public double? ActualPromoNetROIPercent { get; set; }
        public double? ActualPromoNetUpliftPercent { get; set; }
        public double? PlanPromoBaselineBaseTI { get; set; }
        public double? PlanPromoBaseTI { get; set; }
        public double? PlanPromoNetNSV { get; set; }
        public double? ActualPromoBaselineBaseTI { get; set; }
        public double? ActualPromoBaseTI { get; set; }
        public double? ActualPromoNetNSV { get; set; }

        // Promo Closure
        public double? ActualPromoTIShopper { get; set; }
        public double? ActualPromoTIMarketing { get; set; }
        public double? ActualPromoBranding { get; set; }
        public double? ActualPromoBTL { get; set; }
        public double? ActualPromoCostProduction { get; set; }
        public double? ActualPromoCost { get; set; }
        public double? ActualPromoUpliftPercent { get; set; }
        public double? ActualPromoIncrementalLSV { get; set; }
        public double? ActualPromoLSV { get; set; }

        //необходимость полей в таком виде под вопросом
        public double? ActualPromoPostPromoEffectLSV { get; set; }
        public double? ActualPromoPostPromoEffectLSVW1 { get; set; }
        public double? ActualPromoPostPromoEffectLSVW2 { get; set; }

        public double? ActualPromoROIPercent { get; set; }
        public double? ActualPromoIncrementalNSV { get; set; }
        public double? ActualPromoNetIncrementalNSV { get; set; }
        public double? ActualPromoIncrementalMAC { get; set; }

        public bool IsOnInvoice { get; set; }
        public string BrandName { get; set; }
        public string BrandTechName { get; set; }
        public string ProductZREP { get; set; }
        public string PromoStatusName { get; set; }
        public string PromoStatusColor { get; set; }
        public string RejectReasonName { get; set; }
        public string MarsMechanicName { get; set; }
        public string PlanInstoreMechanicName { get; set; }
        public string ActualInStoreMechanicName { get; set; }
        public string MarsMechanicTypeName { get; set; }
        public string PlanInstoreMechanicTypeName { get; set; }
        public string ActualInStoreMechanicTypeName { get; set; }
        public string MechanicComment { get; set; }
        public int? CalendarPriority { get; set; }
        public string Mechanic { get; set; }
        public string MechanicIA { get; set; }
        public double? MarsMechanicDiscount { get; set; }
        public double? PlanInstoreMechanicDiscount { get; set; }
        public string InstoreMechanicName { get; set; }
        public string InstoreMechanicTypeName { get; set; }
        public double? InstoreMechanicDiscount { get; set; }
        public int? Priority { get; set; }
        public string ColorDisplayName { get; set; }
        public string ColorSystemName { get; set; }
        public double DeviationCoefficient { get; set; }

        public bool? IsAutomaticallyApproved { get; set; }
        public bool? IsCMManagerApproved { get; set; }
        public bool? IsDemandPlanningApproved { get; set; }
        public bool? IsDemandFinanceApproved { get; set; }

        public bool? NeedRecountUplift { get; set; }

        public string ProductSubrangesList { get; set; }

        //
        public double? PlanPromoNetIncrementalBaseTI { get; set; }
        public double? PlanPromoNetIncrementalCOGS { get; set; }
        public double? PlanPromoNetBaseTI { get; set; }
        public double? PlanPromoNSV { get; set; }
        public double? ActualPromoLSVByCompensation { get; set; }
        public double? ActualPromoNetIncrementalBaseTI { get; set; }
        public double? ActualPromoNetIncrementalCOGS { get; set; }
        public double? ActualPromoNetBaseTI { get; set; }
        public double? ActualPromoNSV { get; set; }
        public bool? InOut { get; set; }
        // Id операции (по сути транзакции) для предотвращения дублирования
        public Guid? OperationId { get; set; }
        public bool? IsGrowthAcceleration { get; set; }
        public bool? IsApolloExport { get; set; }
        public double? InvoiceTotal { get; set; }
    }
}

File: Module.Frontend.TPM/Client/model/tpm/promo/HistoricalPromo.js
Ext.define('App.model.tpm.promo.HistoricalPromo', {
    extend: 'Ext.data.Model',
    idProperty: '_Id',
    breezeEntityType: 'Promo',
    fields: [
        { name: '_Id', type: 'string', hidden: true },
        { name: '_ObjectId', hidden: true },
        { name: '_User', type: 'string', isDefault: true },
        { name: '_Role', type: 'string', isDefault: true },
        { name: '_EditDate', type: 'date', isDefault: true, timeZone: +3, convert: dateConvertTimeZone},
        { name: '_Operation', type: 'string', isDefault: true },

        { name: 'PromoEventName', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        { name: 'PromoEventDescription', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },

        { name: 'CreatorLogin', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },

        { name: 'IsAutomaticallyApproved', type: "boolean", useNull: true, hidden: true, isDefault: false, defaultValue: null },
        { name: 'IsCMManagerApproved', type: "boolean", useNull: true, hidden: true, isDefault: false, defaultValue: null },
        { name: 'IsDemandPlanningApproved', type: "boolean", useNull: true, hidden: true, isDefault: false, defaultValue: null },
        { name: 'IsDemandFinanceApproved', type: "boolean", useNull: true, hidden: true, isDefault: false, defaultValue: null },

        /***Порядок полей для расширенного фильтра***/
        // Поле из "Basic"
        { name: 'ClientHierarchy', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        // Поле из "Basic"
        { name: 'Name', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        // Поле из "Calculation"
        { name: 'BrandTechName', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        // Поле из "Calculation"
        { name: 'EventName', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        // Поле из "Calculation"
        { name: 'Mechanic', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        { name: 'MarsMechanicName', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        // Поле из "Calculation"
        { name: 'MechanicIA', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        { name: 'PlanInstoreMechanicName', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        // Поле из "Calculation"
        { name: 'PromoStatusName', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        /********************************************/

        { name: 'ProductHierarchy', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'MechanicComment', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'MarsMechanicDiscount', type: 'float', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanInstoreMechanicDiscount', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'StartDate', useNull: true, type: 'date', hidden: false, isDefault: false, defaultValue: null },
        { name: 'EndDate', useNull: true, type: 'date', hidden: false, isDefault: false, defaultValue: null },
        { name: 'DispatchesStart', useNull: true, type: 'date', hidden: false, isDefault: false, defaultValue: null },
        { name: 'DispatchesEnd', useNull: true, type: 'date', hidden: false, isDefault: false, defaultValue: null },
        { name: 'OtherEventName', type: 'string', useNull: true, hidden: true, isDefault: false, defaultValue: null },
        { name: 'CalendarPriority', useNull: true, type: 'int', hidden: false, isDefault: false, defaultValue: null },
        { name: 'NeedRecountUplift', type: 'bool', useNull: true, hidden: true, isDefault: false, defaultValue: null },
        { name: 'InvoiceNumber', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'DocumentNumber', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'IsOnInvoice', type: 'boolean', hidden: false, isDefault: true },

        // Calculation
        { name: 'PlanPromoTIShopper', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoTIMarketing', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoBranding', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoCost', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoBTL', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoCostProduction', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoUpliftPercent', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoIncrementalLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoPostPromoEffectLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoPostPromoEffectLSVW1', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoPostPromoEffectLSVW2', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoROIPercent', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoIncrementalNSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetIncrementalNSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoIncrementalMAC', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoXSites', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoCatalogue', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoPOSMInClient', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoCostProdXSites', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoCostProdCatalogue', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoCostProdPOSMInClient', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoXSites', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoCatalogue', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoPOSMInClient', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoCostProdXSites', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoCostProdCatalogue', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoCostProdPOSMInClient', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoBaselineLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoIncrementalBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoIncrementalCOGS', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoTotalCost', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetIncrementalLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetIncrementalMAC', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoIncrementalEarnings', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetIncrementalEarnings', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetROIPercent', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetUpliftPercent', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoBaselineLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualInStoreDiscount', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualInStoreShelfPrice', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanInStoreShelfPrice', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoIncrementalBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoIncrementalCOGS', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoTotalCost', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetIncrementalLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetIncrementalMAC', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoIncrementalEarnings', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetIncrementalEarnings', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetROIPercent', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetUpliftPercent', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoBaselineBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetNSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoBaselineBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetNSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },

        // Promo Closure
        { name: 'ActualPromoTIShopper', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoTIMarketing', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoBranding', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoBTL', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoCostProduction', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoCost', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoUpliftPercent', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoIncrementalLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoLSVByCompensation', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoPostPromoEffectLSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoPostPromoEffectLSVW1', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoPostPromoEffectLSVW2', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoROIPercent', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoIncrementalNSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetIncrementalNSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoIncrementalMAC', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PromoStatusColor', type: 'string', useNull: true, hidden: true, isDefault: false, defaultValue: null },
        { name: 'MarsMechanicTypeName', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanInstoreMechanicTypeName', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'BrandName', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'ColorSystemName', type: 'string', useNull: true, hidden: true, isDefault: false, defaultValue: null },
        { name: 'ColorDisplayName', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'RejectReasonName', type: 'string', useNull: true, hidden: true, isDefault: false, defaultValue: null },
        { name: 'ActualInStoreMechanicName', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualInStoreMechanicTypeName', type: 'string', useNull: true, hidden: false, isDefault: false, defaultValue: null },

        //
        { name: 'PlanPromoNetIncrementalBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetIncrementalCOGS', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNetBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PlanPromoNSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetIncrementalBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetIncrementalCOGS', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNetBaseTI', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ActualPromoNSV', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },

        // InOut
        { name: 'InOut', useNull: true, type: 'boolean', hidden: false, isDefault: false, defaultValue: null },

        { name: 'LastChangedDate', useNull: true, type: 'date', hidden: false, isDefault: false, defaultValue: null },
        { name: 'LastChangedDateDemand', useNull: true, type: 'date', hidden: false, isDefault: false, defaultValue: null },
        { name: 'LastChangedDateFinance', useNull: true, type: 'date', hidden: false, isDefault: false, defaultValue: null },
        { name: 'PromoDuration', useNull: true, type: 'int', hidden: false, isDefault: false, defaultValue: null },
        { name: 'DispatchDuration', useNull: true, type: 'int', hidden: false, isDefault: false, defaultValue: null },
        { name: 'Number', useNull: true, type: 'int', hidden: false, isDefault: false, defaultValue: null, isKey: true },
        { name: 'Comment', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        { name: 'ProductZREP', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        { name: 'InstoreMechanicName', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        { name: 'InstoreMechanicTypeName', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },
        { name: 'InstoreMechanicDiscount', useNull: true, type: 'float', hidden: false, isDefault: false, defaultValue: null },
        { name: 'Priority', useNull: true, type: 'int', hidden: false, isDefault: false, defaultValue: null },
        { name: 'ProductSubrangesList', type: 'string', useNull: true, hidden: false, isDefault: true, defaultValue: null },

        // Growth Acceleration
        { name: 'IsGrowthAcceleration', useNull: true, type: 'boolean', hidden: false, isDefault: false, defaultValue: null },

        {
            name: 'DeviationCoefficient', type: 'float', hidden: false, isDefault: true,
            convert: function (value) {
                return value * 100;
            }
        },

        //Apollo Export
        { name: 'IsApolloExport', type: 'boolean', hidden: false, isDefault: false }
    ],
    proxy: {
        type: 'breeze',
        resourceName: 'HistoricalPromoes',
        reader: {
            type: 'json',
            totalProperty: 'inlineCount',
            root: 'results'
        },
        extraParams: {
            //Id промо для метода GetById в истории
            promoIdHistory: null
        }
    }
});
